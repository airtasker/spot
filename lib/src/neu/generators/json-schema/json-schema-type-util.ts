import assertNever from "assert-never";
import {
  areBooleanLiteralTypes,
  areFloatLiteralTypes,
  areIntLiteralTypes,
  areStringLiteralTypes,
  isBooleanLiteralType,
  isFloatLiteralType,
  isIntLiteralType,
  isNotLiteralType,
  isStringLiteralType,
  Type,
  TypeKind
} from "../../types";
import { JsonSchemaObject, JsonSchemaType } from "./json-schema-specification";

/**
 * Generate a JSON Schema type definition. `objectAdditionalProperties` may
 * be used to configure whether additional properties should be allowed on
 * object types. This can be useful for data validation purposes where
 * property validation can be strict or lenient.
 *
 * @param type a contract type
 * @param objectAdditionalProperties whether to allow additional properties for objects
 */
export function typeToJsonSchemaType(
  type: Type,
  objectAdditionalProperties: boolean = true // TODO: expose proper configuration object
): JsonSchemaType {
  switch (type.kind) {
    case TypeKind.NULL:
      return {
        type: "null"
      };
    case TypeKind.BOOLEAN:
      return {
        type: "boolean"
      };
    case TypeKind.BOOLEAN_LITERAL:
      return {
        type: "boolean",
        const: type.value
      };
    case TypeKind.DATE:
      return { type: "string", format: "date" };
    case TypeKind.DATE_TIME:
      return { type: "string", format: "date-time" };
    case TypeKind.STRING:
      return {
        type: "string"
      };
    case TypeKind.STRING_LITERAL:
      return {
        type: "string",
        const: type.value
      };
    case TypeKind.FLOAT:
    case TypeKind.DOUBLE:
      return {
        type: "number"
      };
    case TypeKind.FLOAT_LITERAL:
      return {
        type: "number",
        const: type.value
      };
    case TypeKind.INT32:
    case TypeKind.INT64:
      return {
        type: "integer"
      };
    case TypeKind.INT_LITERAL:
      return {
        type: "integer",
        const: type.value
      };
    case TypeKind.OBJECT:
      return type.properties.reduce<JsonSchemaObject & { required: string[] }>(
        (acc, property) => {
          if (!property.optional) {
            acc.required.push(property.name);
          }
          acc.properties[property.name] = typeToJsonSchemaType(
            property.type,
            objectAdditionalProperties
          );
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: [],
          additionalProperties: objectAdditionalProperties
        }
      );
    case TypeKind.ARRAY:
      return {
        type: "array",
        items: typeToJsonSchemaType(
          type.elementType,
          objectAdditionalProperties
        )
      };
    case TypeKind.UNION:
      const elements = type.types;
      if (areBooleanLiteralTypes(elements)) {
        return {
          type: "boolean",
          enum: elements.map(e => e.value)
        };
      } else if (areStringLiteralTypes(elements)) {
        return {
          type: "string",
          enum: elements.map(e => e.value)
        };
      } else if (areFloatLiteralTypes(elements)) {
        return {
          type: "number",
          enum: elements.map(e => e.value)
        };
      } else if (areIntLiteralTypes(elements)) {
        return {
          type: "integer",
          enum: elements.map(e => e.value)
        };
      } else {
        // Guaranteed oneOf

        const oneOfElements = elements
          .filter(isNotLiteralType)
          .map(t => typeToJsonSchemaType(t, objectAdditionalProperties));

        const booleanLiterals = elements.filter(isBooleanLiteralType);
        switch (booleanLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(booleanLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "boolean",
              enum: booleanLiterals.map(b => b.value)
            });
        }

        const stringLiterals = elements.filter(isStringLiteralType);
        switch (stringLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(stringLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "string",
              enum: stringLiterals.map(s => s.value)
            });
        }

        const floatLiterals = elements.filter(isFloatLiteralType);
        switch (floatLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(floatLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "number",
              enum: floatLiterals.map(f => f.value)
            });
        }

        const integerLiterals = elements.filter(isIntLiteralType);
        switch (integerLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(integerLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "integer",
              enum: integerLiterals.map(i => i.value)
            });
        }

        return {
          oneOf: oneOfElements
        };
      }
    case TypeKind.REFERENCE:
      return {
        $ref: `#/definitions/${type.name}`
      };
    default:
      throw assertNever(type);
  }
}
