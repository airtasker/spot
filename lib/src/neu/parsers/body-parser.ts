import { ParameterDeclaration } from "ts-morph";
import { Body } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseType } from "./type-parser";

export function parseBody(
  parameter: ParameterDeclaration,
  typeTable: TypeTable,
  lociTable: LociTable
): Body {
  const decorator = parameter.getDecoratorOrThrow("body");
  if (parameter.hasQuestionToken()) {
    throw new Error("@body parameter cannot be optional");
  }
  const type = parseType(parameter.getTypeNodeOrThrow(), typeTable, lociTable);
  // TODO: add loci information
  return {
    // TODO: how to extract description from parameter declaration?
    description: undefined,
    type
  };
}
