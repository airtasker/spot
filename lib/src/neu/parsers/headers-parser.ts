import { ParameterDeclaration } from "ts-morph";
import { Header } from "../definitions";
import { OptionalNotAllowedError, ParserError } from "../errors";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { err, ok, Result } from "../util";
import {
  getJsDoc,
  getParameterTypeAsTypeLiteralOrThrow,
  getPropertyName
} from "./parser-helpers";
import { parseType } from "./type-parser";

export function parseHeaders(
  parameter: ParameterDeclaration,
  typeTable: TypeTable,
  lociTable: LociTable
): Result<Header[], ParserError> {
  parameter.getDecoratorOrThrow("headers");
  if (parameter.hasQuestionToken()) {
    return err(
      new OptionalNotAllowedError("@headers parameter cannot be optional", {
        file: parameter.getSourceFile().getFilePath(),
        position: parameter.getQuestionTokenNodeOrThrow().getPos()
      })
    );
  }
  const type = getParameterTypeAsTypeLiteralOrThrow(parameter);

  const headers = [];
  for (const propertySignature of type.getProperties()) {
    const typeResult = parseType(
      propertySignature.getTypeNodeOrThrow(),
      typeTable,
      lociTable
    );
    if (typeResult.isErr()) return typeResult;
    const pDescription = getJsDoc(propertySignature);

    const header = {
      name: getPropertyName(propertySignature),
      type: typeResult.unwrap(),
      description: pDescription && pDescription.getComment(),
      optional: propertySignature.hasQuestionToken()
    };
    headers.push(header);
  }

  // TODO: add loci information
  return ok(headers.sort((a, b) => (b.name > a.name ? -1 : 1)));
}
