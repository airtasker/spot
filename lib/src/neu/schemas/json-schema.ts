export interface JsonSchema {
  $schema: "http://json-schema.org/draft-07/schema#";
  definitions: {
    [typeName: string]: JsonSchemaType;
  };
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
  additionalProperties: boolean;
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  format?: string;
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
