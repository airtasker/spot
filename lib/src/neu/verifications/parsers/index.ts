// @ts-ignore
import * as antlr4 from 'antlr4';
// @ts-ignore
import { QueryParamsLexer } from './QueryParamsLexer';
// @ts-ignore
import { QueryParamsParser } from './QueryParamsParser';

// let MyGrammarParser = require('./QueryParamsLexer').MyGrammarParser;
// let MyGrammarListener = require('./QueryParamsLexer').MyGrammarListener;

export function parseQueryParams(input:string) {
    let chars = new antlr4.InputStream(input);
    let lexer = new QueryParamsLexer(chars);
    let tokens  = new antlr4.CommonTokenStream(lexer);
    // console.log(tokens);
    let parser = new QueryParamsParser(tokens);
    parser.buildParseTrees = true;
    let tree = parser.querystring();
    console.log(tree.toStringTree(parser.ruleNames)); 
}
