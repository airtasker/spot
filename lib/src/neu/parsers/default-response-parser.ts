import { MethodDeclaration } from "ts-morph";
import { DefaultResponse } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseBody } from "./body-parser";
import { parseHeaders } from "./headers-parser";
import { getJsDoc, getParamWithDecorator } from "./parser-helpers";

export function parseDefaultResponse(
  method: MethodDeclaration,
  typeTable: TypeTable,
  lociTable: LociTable
): DefaultResponse {
  method.getDecoratorOrThrow("defaultResponse");
  const headersParam = getParamWithDecorator(method, "headers");
  const headers = headersParam
    ? parseHeaders(headersParam, typeTable, lociTable)
    : [];
  const bodyParam = getParamWithDecorator(method, "body");
  const body = bodyParam && parseBody(bodyParam, typeTable, lociTable);
  const descriptionDoc = getJsDoc(method);
  return {
    headers,
    description: descriptionDoc && descriptionDoc.getComment(),
    body
  };
}
