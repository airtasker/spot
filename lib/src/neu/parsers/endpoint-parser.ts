import { ClassDeclaration, TypeGuards } from "ts-morph";
import { EndpointConfig } from "../../syntax/endpoint";
import { Endpoint } from "../definitions";
import { ParserError } from "../errors";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { err, ok, Result } from "../util";
import { parseDefaultResponse } from "./default-response-parser";
import {
  getDecoratorConfigOrThrow,
  getJsDoc,
  getMethodWithDecorator,
  getObjLiteralProp,
  getObjLiteralPropOrThrow,
  getPropValueAsArrayOrThrow,
  getPropValueAsStringOrThrow,
  isHttpMethod
} from "./parser-helpers";
import { parseRequest } from "./request-parser";
import { parseResponse } from "./response-parser";

export function parseEndpoint(
  klass: ClassDeclaration,
  typeTable: TypeTable,
  lociTable: LociTable
): Result<Endpoint, ParserError> {
  const decorator = klass.getDecoratorOrThrow("endpoint");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);

  // Handle name
  const endpointName = klass.getNameOrThrow();

  // Handle method
  const methodProp = getObjLiteralPropOrThrow<EndpointConfig>(
    decoratorConfig,
    "method"
  );
  const methodLiteral = getPropValueAsStringOrThrow(methodProp);
  const methodValue = methodLiteral.getLiteralText();
  if (!isHttpMethod(methodValue)) {
    throw new Error(`expected a HttpMethod, got ${methodValue}`);
  }

  // Handle path
  const pathProp = getObjLiteralPropOrThrow<EndpointConfig>(
    decoratorConfig,
    "path"
  );
  const pathLiteral = getPropValueAsStringOrThrow(pathProp);

  // Handle tags
  const tagsProp = getObjLiteralProp<EndpointConfig>(decoratorConfig, "tags");
  const tagsLiteral = tagsProp && getPropValueAsArrayOrThrow(tagsProp);
  const tagsValueLiterals = [];
  if (tagsLiteral) {
    for (const elementExpr of tagsLiteral.getElements()) {
      if (TypeGuards.isStringLiteral(elementExpr)) {
        tagsValueLiterals.push(elementExpr);
      } else {
        return err(
          new ParserError("tag must be a string", {
            file: elementExpr.getSourceFile().getFilePath(),
            position: elementExpr.getPos()
          })
        );
      }
    }
  }

  // Handle jsdoc
  const descriptionDoc = getJsDoc(klass);

  // Handle request
  const requestMethod = getMethodWithDecorator(klass, "request");
  let request;
  if (requestMethod) {
    const requestResult = parseRequest(requestMethod, typeTable, lociTable, {
      endpointName
    });
    if (requestResult.isErr()) return requestResult;
    request = requestResult.unwrap();
  }

  // Handle responses
  const responseMethods = klass
    .getMethods()
    .filter(m => m.getDecorator("response") !== undefined);
  const responses = [];
  for (const method of responseMethods) {
    const responseResult = parseResponse(method, typeTable, lociTable);
    if (responseResult.isErr()) return responseResult;
    responses.push(responseResult.unwrap());
  }
  // TODO: find duplicate response statuses

  // Handle default response
  const defaultResponseMethod = getMethodWithDecorator(
    klass,
    "defaultResponse"
  );
  let defaultResponse;
  if (defaultResponseMethod) {
    const defaultResponseResult = parseDefaultResponse(
      defaultResponseMethod,
      typeTable,
      lociTable
    );
    if (defaultResponseResult.isErr()) return defaultResponseResult;
    defaultResponse = defaultResponseResult.unwrap();
  }

  // Add location data
  lociTable.addMorphNode(LociTable.endpointClassKey(endpointName), klass);
  lociTable.addMorphNode(
    LociTable.endpointDecoratorKey(endpointName),
    decorator
  );
  lociTable.addMorphNode(LociTable.endpointMethodKey(endpointName), methodProp);
  lociTable.addMorphNode(LociTable.endpointPathKey(endpointName), pathProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(
      LociTable.endpointDescriptionKey(endpointName),
      descriptionDoc
    );
  }
  if (tagsProp) {
    lociTable.addMorphNode(LociTable.endpointTagsKey(endpointName), tagsProp);
    tagsValueLiterals.forEach(tagLiteral => {
      lociTable.addMorphNode(
        LociTable.endpointTagKey(endpointName, tagLiteral.getLiteralText()),
        tagLiteral
      );
    });
  }
  return ok({
    name: endpointName,
    description: descriptionDoc && descriptionDoc.getComment(),
    tags: tagsValueLiterals.map(tvl => tvl.getLiteralText()), // TODO: sort tags
    method: methodValue,
    path: pathLiteral.getLiteralText(),
    request,
    responses: responses.sort((a, b) => (b.status > a.status ? -1 : 1)),
    defaultResponse
  });
}
