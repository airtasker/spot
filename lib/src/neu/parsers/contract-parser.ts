import { ClassDeclaration, SourceFile } from "ts-morph";
import { ApiConfig } from "../../syntax/api";
import { Config, Contract } from "../definitions";
import { ParserError } from "../errors";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { ok, Result } from "../util";
import { defaultConfig, parseConfig } from "./config-parser";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

/**
 * Parse a root source file to return a contract.
 */
export function parseContract(
  file: SourceFile
): Result<{ contract: Contract; lociTable: LociTable }, Error> {
  const typeTable = new TypeTable();
  const lociTable = new LociTable();

  const klass = getClassWithDecoratorOrThrow(file, "api"); // TODO: throw a custom error
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const descriptionDoc = getJsDoc(klass);

  // Handle config
  const configResult = resolveConfig(klass);
  if (configResult.isErr()) return configResult;

  // Handle security
  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");
  const security =
    securityHeaderProp &&
    parseSecurityHeader(securityHeaderProp, typeTable, lociTable);
  if (security && security.isErr()) return security;

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpointClasses = projectFiles.reduce<ClassDeclaration[]>(
    (acc, currentFile) =>
      acc.concat(
        currentFile
          .getClasses()
          .filter(k => k.getDecorator("endpoint") !== undefined)
      ),
    []
  );
  const endpoints = [];
  for (const k of endpointClasses) {
    const endpointResult = parseEndpoint(k, typeTable, lociTable);
    if (endpointResult.isErr()) return endpointResult;
    endpoints.push(endpointResult.unwrap());
  }

  const contract = {
    name: nameLiteral.getLiteralText(),
    description: descriptionDoc && descriptionDoc.getComment(),
    types: typeTable.toArray(),
    config: configResult.unwrap(),
    security: security && security.unwrap(),
    endpoints: endpoints.sort((a, b) => (b.name > a.name ? -1 : 1))
  };

  return ok({ contract, lociTable });
}

function resolveConfig(klass: ClassDeclaration): Result<Config, ParserError> {
  const hasConfigDecorator = klass.getDecorator("config") !== undefined;
  if (hasConfigDecorator) {
    return parseConfig(klass);
  } else {
    return ok(defaultConfig());
  }
}
