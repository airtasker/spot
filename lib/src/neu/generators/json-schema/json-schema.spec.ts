import { isJSONSchemaDraft7, Spectral } from "@stoplight/spectral";
import { createProjectFromExistingSourceFile } from "../../../spec-helpers/helper";
import { parseContract } from "../../parsers/contract-parser";
import { generateJsonSchema } from "./json-schema";

describe("JSON Schema generator", () => {
  const spectral = new Spectral();

  beforeAll(async () => {
    spectral.registerFormat("json-schema-draft7", isJSONSchemaDraft7);
  });

  test("produces minimal json schema", async () => {
    const file = createProjectFromExistingSourceFile(
      `${__dirname}/__spec-examples__/minimal-contract.ts`
    ).file;

    const { contract } = parseContract(file).unwrapOrThrow();
    const result = generateJsonSchema(contract);

    expect(result.$schema).toEqual("http://json-schema.org/draft-07/schema#");
    expect(result.definitions).toEqual({});
    expect(result).toMatchSnapshot();
    const spectralResult = await spectral.run(result);
    expect(spectralResult).toHaveLength(0);
  });

  test("produces definitions", async () => {
    const file = createProjectFromExistingSourceFile(
      `${__dirname}/__spec-examples__/contract-with-reference-types.ts`
    ).file;

    const { contract } = parseContract(file).unwrapOrThrow();
    const result = generateJsonSchema(contract);

    expect(result.definitions).toHaveProperty("User");
    expect(result.definitions).toHaveProperty("Users");
    expect(result).toMatchSnapshot();
    const spectralResult = await spectral.run(result);
    expect(spectralResult).toHaveLength(0);
  });
});
