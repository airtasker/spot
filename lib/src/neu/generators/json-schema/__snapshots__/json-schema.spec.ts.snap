// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Schema generator produces definitions 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {
    "User": Object {
      "additionalProperties": true,
      "properties": Object {
        "id": Object {
          "type": "string",
        },
        "name": Object {
          "type": "string",
        },
      },
      "required": Array [
        "id",
        "name",
      ],
      "type": "object",
    },
    "Users": Object {
      "items": Object {
        "$ref": "#/definitions/User",
      },
      "type": "array",
    },
  },
}
`;

exports[`JSON Schema generator produces minimal json schema 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": Object {},
}
`;
